Bug 797540 - Mapping of aqbanking accounts to gnucash accounts doesn'…

pkzw opened the pull request:
This pull request fixes issues with the AqBanking/GnuCash account matching:

- After leaving the Online Banking Setup wizard and entering the wizard again the previously maintained matches were no longer displayed.

- If an AqBanking account (ab_acc) was matched to a different GnuCash account (gnc_acc) the old match was not removed resulting in an AqBanking account matched to two GnuCash accounts.

The issues are fixed by

- introducing a dedicated callback function to test two AqBanking accounts for equality and an associated hash function returing the account unique id as hash value.

- using this equality check also for checking if a newly assigned GnuCash account has been previously been assigned to a different AqBanking account and deleting the previous assignment to keep the matches unique.

- additionally, a reverse hash table was introduced in order to memorized the matched GnuCash accounts before starting the matching wizard. This hash table is used within the aai_on_finish callback to delete only the KVPs of those GnuCash accounts which are no longer matched with an AqBanking account. All other GnuCash accounts (previously matched and currently matched) are just updated with the new assigments.

jralls commented:
I see that you've figured out how to make a PR. Please break this up into 3 commits, each doing one of the bullet points under "The issues are fixed by". Each commit message should include a description of what it does; the text of those bullet points will work well.

pkzw commented on gnucash/import-export/aqb/assistant-ab-initial.c:
```diff
@@ -290,6 +305,49 @@ aai_wizard_button_clicked_cb(GtkButton *button, gpointer user_data)
+       GNC_AB_ACCOUNT_SPEC *ab_acc = (GNC_AB_ACCOUNT_SPEC *) key;
+       GHashTable *accounts = (GHashTable *) user_data;
+       g_return_if_fail(accounts && value && ab_acc);
+       g_hash_table_insert(accounts, value, (gpointer) ab_acc);
```
OK, I will remove it.

jralls replied:
Since as you say below that you only need revhash's key then store it in both key and value, g_hash_table_insert recognizes and optimizes for that special case.

pkzw commented on gnucash/import-export/aqb/assistant-ab-initial.c:
```diff
@@ -290,6 +305,49 @@ aai_wizard_button_clicked_cb(GtkButton *button, gpointer user_data)
+{
+       GHashTable *accounts = (GHashTable *) user_data;
+       g_return_if_fail(accounts && value);
+       g_hash_table_remove(accounts, value);
```
info->revhash is just needed to store (GnuCash account) keys. The value (of the hash table entry) is not needed. The value parameter of this callback function is the value of the other hash table (i.e. info->hash), which is also a GnuCash account.
I will also remove local variable accounts.

jralls replied:
To make the intent more clear then rename this function `remove_revhash_cb`, rename `user_data` to `revhash`, and add a comment that `value` in `info->hash` is the key for `info->revhash`.

pkzw commented on gnucash/import-export/aqb/assistant-ab-initial.c:
```diff
@@ -332,11 +396,14 @@ aai_on_finish (GtkAssistant *assistant, gpointer user_data)
+    /* Remove GnuCash accounts from reverse hash table which are still
+     * matched to an AqBanking account. For the remaining GnuCash accounts
+     * the KVPs must be cleared (i.e. deleted) */
+    g_hash_table_foreach(info->gnc_hash, (GHFunc) remove_acc_cb, info->gnc_revhash);
```
OK. looks better :-)

pkzw replied:
After rethinking your suggestion I come to the conclusion that it's not equivalent to my implementation. Let's make the following example:

On entrance to the account matching page `info->hash` matching AqBanking accounts (ab_acc) with GnuCash accounts (gnc_acc) may look like

ab_acc | gnc_acc
--- | ---
a1 | -
a2 | g2
a3 | g1
a4 | g3

This would result in `info->revhash` look like

gnc_acc | ab_acc
--- | ---
g1  | a3
g2  | a2
g3  | a4

Let's assume that the user now will match g1 with a1, g4 with a4 (please note that a3 is unmatched after the first operation). This would finally result in the following `info->hash` on exit of the account matching page:

ab_acc | gnc_acc
--- | ---
a1 | g1
a2 | g2
a3 | -
a4 | g4

As a result of this g1 and g2 shall be removed from `info->revhash`:

gnc_acc | ab_acc
--- | ---
~~g1~~  | ~~a3~~
~~g2~~  | ~~a2~~
g3  | a4

That means that there is only g3 remaining in `info->revhash`, which is now no longer matched with an AqBanking account and whose KVPs need to be deleted.

As far as I can see your suggestion will iterate through `info->revhash` and will lookup matches in `info->hash` by the AqBanking accounts a3, a2, and a4. The callback function `in_forward_hash` will return `TRUE` only for a2 and a4 and thus will cause a removal of g2 and g3 from `info->revhash`, and this will cause deletion of the KVPs for g1 but will keep the KVPs for g3.

I think I will keep the code as it is and may add further comments. Otherwise `in_forward_hash` should lookup `info->hash` by value, which seems to be more inefficient to me.

jralls replied:
OIC. That can be corrected by changing the condition in `in_forward_hash` to
`if (gnc_hash_lookup_value((GHashTable*)user_data, value) == key)` *but* this approach depends on *not* using the `g_hash_table_insert` optimization of passing `key` to both `key` and `value` because this way you do need `ab_acc` in `revhash`.

jralls replied:
Strike that, checking that `g_hash_lookup((GHashTable*)user_data, value) == key` would also exclude g1 as that's changed. `remove_acc_cb` modified as I suggest is indeed the right way.

pkzw commented:
OK, I could break it into 3 Commits. Let's call them C1, C2 and C3. C3
will be independent of C1 and C2 (just one changed line) and C2 will
depend on C1. Shall I create two (more) branches for commit C1+C2 and
for C3, respectively, and create a pull request for each branch?
Or shall I revert my previous commit and create C1, C2, and C3 on the
old branch?
Please advice.

jralls commented:
You can simply `git reset HEAD^` locally to remove the commit while leaving the code unchanged, then do each commit in turn, and finally force-push the result to Github. The PR is based on the branch so anything that you do in the branch is magically carried into the PR.

pkzw commented:
Ok, this is the 2nd iteration of the PR. I have now split the changes into 3 commits as recommended.
I have also renamed the callback functions for the insertions and removals from the reverse hash table and have added a comment. Beside this I have also done a few more cosmetic changes.

For the remaining issues/changes I will create two new bugzillas and I will create a new PR for each. OK?

jralls commented:
No need for bug reports, just make PRs.

jralls commented:
The code is much clearer, thanks.

The commit messages need a bit of work. They should consist of a summary of less than 80 characters followed by any detail you want to add. One of them should have the bug summary just like the PR title; I'd think the second one is right for that. Otherwise the summary should explain what the commit does, e.g.
"Clear hbci KVP from accounts no longer associated with an online one." for the third commit.

If you're committing from the command line rather than a tool you can use multiple `-m` arguments to `git commit`. The first will be the summary, subsequent ones are added separated by a blank line. For longer messages like the third one it's more convenient to use no `-m` argument so that `git commit` will launch your $EDITOR to allow you to enter the summary and discussion, just separate the summary from the rest with a blank line. You can use `git rebase -i` to edit each of the existing commit messages in turn; there are further instructions in the rebase file that git will open in your $EDITOR.

pkzw commented:
I hope the reworded commits look better now.

jralls commented:
They do, except you didn't change the middle one to the bug summary. That's important for release notes so I've done it as part of merging. Thanks for persisting.

pkzw commented:
Hi John, thanks for your assistance. There are two issues/corrections open from my patch file prior to this PR. I think, I will only delete the branch but keep the fork for two more pull requests.
